core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -358,10 +358,20 @@ public long spill(long numBytes) throws IOException {
           return 0L;
         }
 
+        long released = 0L;
+        // longArray will not be used anymore, but may take significant amount of memory
+        // but never released until the task is complete.
+        if (longArray != null) {
+          released += longArray.memoryBlock().size();
+          freeArray(longArray);
+          longArray = null;
+        }
+        if (released >= numBytes) {
+          return released;
```

Comment:
@cloud-fan, I was wondering this, too: I think that this is okay because we're in a destructive iterator over the map, so we can just walk through the data pages without needing to worry about hash-style lookups. If this is the case, though, then we should just destroy the pointer array earlier (as I commented downthread on the PR).